### PR TITLE
feat: add OpenAI fdw

### PR DESCRIPTION
**Pull request by burmecia**
## What kind of change does this PR introduce?

This PR is to add an OpenAI foreign data wrapper to fetch embedding data for input. This FDW supports reading embedding data based on [text-embedding-ada-002 model](https://platform.openai.com/docs/guides/embeddings/second-generation-models), the embedding will be in `string` format which can be saved to a `vector` column created by [pgvector](https://github.com/pgvector/pgvector).

## What is the current behavior?

There is no OpenAI foreign data wrapper.

## What is the new behavior?

The new OpenAI foreign data wrapper will be added.

## Additional context

There is no mock OpenAI api available yet, so test cases are not added.

### CODE REVIEW

File: docs/openai.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
+[OpenAI](https://openai.com/) is an American artificial intelligence research laboratory consisting of the non-profit OpenAI Incorporated and its for-profit subsidiary corporation OpenAI Limited Partnership. OpenAI conducts AI research with the declared intention of promoting and developing a friendly AI.
```
Comment (kiwicopple) [inline docs/openai.md:1]:
```suggestion
[OpenAI](https://openai.com/) is an artificial intelligence research laboratory that conducts AI research with the intention of promoting and developing a friendly AI.
```

your version sounds a bit politicized 😅

File: docs/openai.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+```sql
+-- save OpenAI API key in Vault and get its key id
+select pgsodium.create_key(name := 'openai');
+insert into vault.secrets (secret, key_id) values (
+  'sk-xxx',
+  (select id from pgsodium.valid_key where name = 'openai')
+) returning key_id;
+```
+
+Then create the foreign server,
```
Comment (kiwicopple) [inline docs/openai.md:36]:
```suggestion
Then create the foreign server:
```

File: docs/openai.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  foreign data wrapper openai_wrapper
+  options (
+    api_key 'sk-xxx'
+  );
+```
+
+
+### Tables
+
+OpenAI wrapper currently only supports reading [embedding](https://platform.openai.com/docs/guides/embeddings/what-are-embeddings) data through OpenAI API. The foreign table must contain `id` and `embedding` column.
```
Comment (kiwicopple) [inline docs/openai.md:75]:
```suggestion
OpenAI wrapper currently supports reading [embedding](https://platform.openai.com/docs/guides/embeddings/what-are-embeddings) data through OpenAI API. The foreign table must contain `id` and `embedding` column.
```

File: docs/openai.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        source '(select id, input from my_source where column1='${_param1}' and column2=${_param2})'
+      );
+
+    select * from my_table where _param1='aaa' and _param2='32';
+   ```
+
+
+#### Examples
+
+A simple example,
```
Comment (kiwicopple) [inline docs/openai.md:115]:
```suggestion
A simple example:
```

File: docs/openai.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  embedding = s.embedding::vector,
+  ts = now()
+from my_embedding2 s
+where s.id = t.id
+  and s._param_author = 'foo'
+  and s._param_ts = '2023-05-21';
+
+-- check if embedding has been saved
+select * from my_table2;
+```
```
Comment (kiwicopple) [inline docs/openai.md:208]: Brainstorming more examples which will help with the DX:

## Example 3

This one allows the user to specify a limit for how many embeddings to fetch at once:

```sql
create table documents (
  id serial primary key,
  content text not null,
  embedding vector(1536)
);

create foreign table documents_embeddings (
  id bigint,
  embedding text,
  num_rows smallint
)
  server openai_server
  options (
    -- Note: the source sub-query must have an 'id' and an 'input' column, and the corresponding parameters.
    source '(
      select id, content as input
      from documents
      limit ${num_rows}
    )'
  );

```

File: docs/openai.md (hunk 1 above)
Comment (burmecia) [reply docs/openai.md:1]: haha, that is what I copied from Wikipedia. 😄

**Comment by burmecia** [comment]
close for not active for long time.

**Comment by Heniballer** [comment]
Can I still fit in

**Pull request closed without merging**